### ISSUE TITLE
[Feature Request] Mid-Frame Pnach Patches/Cheats

### ISSUE THREAD

**Issue by ZwipZwapZapony** [PCSX2/pcsx2#3327]
_PCSX2 version, although not applicable: v1.5.0-dev-3400-g56b76b17a_

\
PCSX2's built-in patch/cheat system has two pnach patch "placements", being once at the boot-up of the game (0) and once at every V-blank/every frame (1). However, sometimes that's not enough.

For example, in Ratchet: Deadlocked (NTSC-U) [SCUS-97465], let's try using this code...
`patch=1,EE,2036BF20,word,40000000`
This code sets player 1's movement speed to 200% _(edit: possibly only while in DreadZone, including the hub area, not any of the planets)_, and I've used Cheat Engine to verify that this address affects the player both while grounded and while airborne. (I verified that by using Cheat Engine to NOP out anything writing to the address, giving me full control over its value.)

\
However, when just using that patch in PCSX2 without something like Cheat Engine, the game overwrites that value with the default 100% movement speed every frame (while in singleplayer).
Due to the orders of operations, that patch thusly seems to do one of these things every frame:

\- If the player is grounded, PCSX2's patch kicks in (setting the speed to 200%), then the game checks the value (making the player move at 200% speed), then the game overwrites the value (setting the speed to 100%).
As such, the player _can_ move super fast while grounded.

\- If the player is airborne, PCSX2's patch kicks in (setting the speed to 200%), then the game overwrites the value (setting the speed to 100%), then the game checks the value (making the player move at 100% speed).
As such, the player **_cannot_** move super fast while airborne.

\
This means that the player moves at 200% speed while grounded, but only 100% while airborne.
Again, I've confirmed using Cheat Engine that that address affects both grounded airborne speeds, and that it gets overwritten by the game every frame.

And it seems that whatever rewrites the default does so from a different address every time that a new level is loaded, so I can't just NOP out the rewriting from within a pnach file (at least not easily).

Therefore, I'd like to propose a "mid-frame" patch placement.
It should either make PCSX2 check when the address is written to, and then immediately write over it afterwards with the patch value, or prevent the emulation from writing to the address in the first place, or make it so that anything reading the address will simply read the patch value instead.
The second of those options sounds "best" to me, although I won't claim to know what's truly actually best or feasible.

\
\
**TL;DR:** I want to patch an address, but the game overwrites that address mid-frame with the default value before reading it, so I'd like to suggest a new patch placement that can handle that sort of situation.

**Comment by RazielZnot** [comment]
You can't know when exactly game updates value in the RAM. Such feature is totally useless, because different games have different internal loops. Use debugger to find which code updates this value and patch it appropriately. You need to know MIPS assembly to do it properly, because games can call same function from different parts of the code. When you find the code responsible for updating this value and patch it properly, it won't matter anymore when this patch will be applied.

Also using Cheat Engine for editing game's code is not safe, because you are editing recompiled code, it can be recompiled anytime and your changes will be destroyed. Cheat Engine is made specifically for x86 code, it can't handle MIPS code at all, always use PCSX2's debugger for this.

**Comment by RedPanda4552** [comment]
Agreed. Basically what has been described here is similar to that of a memory breakpoint, which is a feature of the current PCSX2 debugger. Though there is potential for implementing something like this, I think the much more reasonable solution is to instead change the patch to prevent the constant writes to the memory location. I wouldn't say this issue is invalid, but I don't really see much of a case for supporting this unless we were to find a moderate to significant number of games that somehow have this problem.

**Comment by lightningterror** [comment]
Closing, trivial.
